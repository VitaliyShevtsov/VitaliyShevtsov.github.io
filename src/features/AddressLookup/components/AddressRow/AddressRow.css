.list-item-marker {
  display: flex;
  align-items: center;
  justify-content: center;

  height: 40px;
  width: 40px;
  min-width: 40px;
  min-height: 40px;
  border-radius: 20px;

  color: rgb(39, 39, 42);
  font-weight: 500;
  font-size: 1rem;
  background: rgb(228, 228, 231);
}
