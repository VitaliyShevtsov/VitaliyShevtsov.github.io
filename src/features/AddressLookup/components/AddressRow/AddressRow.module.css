.listItem {
  margin: 1rem 0;
}

.listItemMarker {
  display: flex;
  align-items: center;
  justify-content: center;

  height: 34px;
  width: 34px;
  min-width: 34px;
  min-height: 34px;
  border-radius: 16px;

  color: rgb(39, 39, 42);
  font-weight: 500;
  font-size: 1rem;
  background: rgb(228, 228, 231);
}
