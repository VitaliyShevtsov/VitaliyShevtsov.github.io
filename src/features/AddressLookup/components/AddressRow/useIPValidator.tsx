
interface ValidationResult {
  readonly valid: boolean;
  readonly message?: string;
}

export function useAddressValidate(): (value: string) => ValidationResult {
  const validator = (value: string) => {
    const parts = value.split(/[.:]/);

    const validator = ipValidator();

    console.log(validator('2001:0db8:85a3:0000:0000:8a2e:0370:7334'))

    if (parts.length === 4) {
      for (const part of parts) {
        const num = parseInt(part);
        if (isNaN(num) || num < 0 || num > 255) {
          return { valid: false };
        }
      }
      return { valid: true };
    } else if (parts.length === 8) {
      for (const part of parts) {
        if (!/^[0-9a-fA-F]{1,4}$/.test(part)) {
          return { valid: false };
        }
      }
      return { valid: true };
    }

    return { valid: false };
  };

  return validator;
}


