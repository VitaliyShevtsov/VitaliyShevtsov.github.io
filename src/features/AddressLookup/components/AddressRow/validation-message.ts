export const ValidationMessage = {
  invalidIP: 'Invalid IP address',
  blankValue: 'Value cannot be blank',
} as const;
