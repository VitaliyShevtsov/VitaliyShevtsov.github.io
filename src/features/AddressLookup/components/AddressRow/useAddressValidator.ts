import isMyIpValid from 'is-my-ip-valid';
import { ValidationMessage } from './validation-message';

type ValidationResult = { valid: boolean; message?: string };

export function useAddressValidator() {
  const validator = (value: string): ValidationResult => {
    const validate = isMyIpValid();

    if (!value) {
      return { valid: false, message: ValidationMessage.blankValue };
    }

    if (validate(value)) {
      return { valid: true };
    }

    return { valid: false, message: ValidationMessage.invalidIP };
  };

  return validator;
}
