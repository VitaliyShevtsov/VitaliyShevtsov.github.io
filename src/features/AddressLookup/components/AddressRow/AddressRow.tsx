import { memo, useCallback, useState, type FocusEventHandler } from 'react';
import type { RecordRow } from '../../types';
import { Box, Field, Image, Input, Spinner, Stack } from '@chakra-ui/react';
import { Tooltip } from '@/components/ui/tooltip';
import AddressClock from '../AddressClock';
import styles from './AddressRow.module.css';

import { useAddressValidator } from './useAddressValidator';
import { DataTestId } from '@/test';

interface Props {
  readonly rowNum: number;
  readonly row: RecordRow;
  readonly fetchAddress: (ip: string, id: string) => void;
  readonly clearRow: (id: string) => void;
}

const AddressRow: React.FC<Props> = ({ rowNum, row, fetchAddress, clearRow }) => {
  const [errorMessage, setErrorMessage] = useState<string | null>(null);
  const validate = useAddressValidator();

  const handleInputBlur: FocusEventHandler<HTMLInputElement> = useCallback(
    (event) => {
      const value = event.target.value.trim();

      if (value === row.ip) {
        return;
      }

      clearRow(row.id);
      const { valid, message } = validate(value);

      if (valid) {
        setErrorMessage(null);
        fetchAddress(value, row.id);
      } else {
        setErrorMessage(message!);
      }
    },
    [fetchAddress, validate, clearRow, row.id, row.ip]
  );

  return (
    <li className={styles.listItem}>
      <Stack direction="row" alignItems="center" paddingBottom="14px">
        <Box className={styles.listItemMarker}>{rowNum}</Box>

        <Field.Root flex={'1 1 70%'} invalid={Boolean(errorMessage)}>
          <Input
            disabled={row.loading}
            type="text"
            placeholder="0.0.0.0"
            data-testid={DataTestId.AddressInput}
            onBlur={handleInputBlur}
          />
          <Field.ErrorText position="absolute" bottom="-20px">
            {errorMessage}
          </Field.ErrorText>
        </Field.Root>

        <Box flex={'0 1 110px'}>
          {row.loading ? <Spinner size="md" data-testid={DataTestId.AddressRowSpinner} /> : null}
          {!row.loading && row.record ? (
            <Stack direction="row" alignItems="center" data-testid={DataTestId.AddressInfoStack}>
              <Tooltip content={`${row.record.city}, ${row.record.country}`}>
                <Image cursor="pointer" boxShadow="sm" height="30px" src={row.record.flag.img} />
              </Tooltip>

              <AddressClock timezone={row.record.timezone} />
            </Stack>
          ) : null}
        </Box>
      </Stack>
    </li>
  );
};

export default memo(AddressRow);
