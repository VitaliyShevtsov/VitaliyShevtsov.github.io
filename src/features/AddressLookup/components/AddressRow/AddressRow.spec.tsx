import { componentTestBed } from '@/test';

import { describe, expect, it } from 'vitest';
import AddressRow from './AddressRow';

describe('<AddressRow />', () => {
  it('should match snapshot', () => {
    const rowNum = 1;
    const row = { id: 1, loading: false } as const;
    const fetchAddress = vi.fn(() => undefined);
    const clearRow = vi.fn(() => undefined);

    const props = { rowNum, row, fetchAddress, clearRow };

    const { container } = componentTestBed(<AddressRow {...props} />);

    expect(container).toMatchSnapshot();
  });
});
