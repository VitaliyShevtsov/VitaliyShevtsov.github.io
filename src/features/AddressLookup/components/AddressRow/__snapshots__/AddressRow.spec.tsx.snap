// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AddressRow /> > should match snapshot 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <li
    class="_listItem_06f2f2"
  >
    <div
      class="chakra-stack css-12433jz"
    >
      <div
        class="_listItemMarker_06f2f2 css-0"
      >
        1
      </div>
      <div
        class="chakra-field__root css-1wgh383"
        data-part="root"
        data-scope="field"
        id="field::«r0»"
        role="group"
      >
        <input
          class="chakra-input css-ee9xfg"
          data-part="input"
          data-scope="field"
          id="«r0»"
          placeholder="0.0.0.0"
          type="text"
        />
      </div>
      <div
        class="css-1dmnsst"
      />
    </div>
  </li>
   
</div>
`;
